Mount Data and Metadata on provided partition with systemd + Best practices (#59)
YunoHost-Apps/garage_ynh: doc/ADMIN.md

# Limitations

  * Garage can be used with only one or two servers (=replica), but requires a cluster with at least 3 nodes to provide its full replication potential (data loss prevention). See [replication modes](https://garagehq.deuxfleurs.fr/documentation/reference-manual/configuration/#replication-mode). It is mainly designed to be used with 3 nodes or more.
  * If you are behind a NAT and use UPnP to configure your port redirection, you may need to
  * add peers via config panel instead of during installation
  * and/or create a permanent port redirection in your router/box

# Informations you should be aware of:

> * This application provides a Node that you can connect to a Garage Cluster. The Config Panel provides some Node-related settings. Cluster-related settings like Key or Bucket should be managed via Garage CLI or by another Node

Reviewer: We have a UI coming soon that could allow you do those operations, and there are at least two community ones
- https://github.com/Noooste/garage-ui
- https://github.com/khairul169/garage-webui

Author: OK then we will not spend much time working on the garage_ynh config panel. Will it be integrated in garage or will need to be installed as a separate package ?

Author: https://github.com/YunoHost-Apps/garage_ynh/issues/62

Reviewer: The two community ones need to be setup as a separate program, the UI that we plan will be embedded into garage

* Garage does not set a hard storage Quota. So disk space usage can become greater than the Node Weight you define. 
* To prevent this, this App will ask you for a Data partition at install, or create a virtual disk with size the Weight of your Node. Virtual disk is only created if YunoHost is not in a VM, and performance may be poor.
* If you install on a VM and do not provide a Data partition, you are responsible for checking sufficient available storage.

> * Expect Metadata (database) and its 2-3 snapshots to be approximatly 3% of the Data size (3GB for 100GB of Data for instance), or a bigger percentage if you store many small objects.

Reviewer: Objects below 3kb will be stored directly in the metadata database

> * To connect from another Node, you might need the RPC port. Is is defined in `rpc_bind_addr` in your `garage.toml`

Reviewer: ```suggestion
* To connect from another Node, you will need the RPC port. Is is defined in `rpc_bind_addr` in your `garage.toml`
```

## How to run CLI commands for Garage

* Type `yunohost app shell garage` on the server command line to enter Garage environment. You will be located in __INSTALL_DIR__ Garage directory.
* To run a command, you need to specify the config file as a parameter `garage -c garage.toml [the actions you wish to run]`.
* Don't forget to `exit` the environment when you're done.

# Configuration depending on your server setup

We distinguish between 2 major setups:
* Self-hosting-like
  * Power (micro-)outage or cat-unplug likely: Metadata are likely to get corrupt
  * Networking can be slower or have poor level of service
  * Data storage up to some TB
  * Usage for distributed high-capacity storage, e.g. backups. Not necessarily for performance
  * Usage for non-critical services with ephemeral data
  * Recovery from other Nodes within a day (no Data snapshot)
* Data-center-like
  * Unclean shutdown unlikely
  * Fast and high-service networking
  * Data storage greater than some TB
  * Usage for high-performance distributed storage
  * Local Recovery of Node within minutes, e.g. with Data snapshot

**Recommended** (minimal) self-hosting config:

> * Data partition: on **SSD** (HDD OK if no high-performance storage), **XFS**, BTRFS & ZFS OK too (EXT4 only for storage in the GB order)

Reviewer: ```suggestion
* Data partition: on **SSD** (HDD OK if no high-performance storage), **XFS**, BTRFS & ZFS OK too (EXT4 can limit you due to the restricted amount of inodes available)
```

Author: Suppose yunohost users don't now what inodes are. What is the inode limit ? 
I prefer to give hints that can help users decide depending on their use-case.  Like a rough number of files and number of GB above which not to use ext4. 1, 10, 100, 1000 GB ? 1 10TB ?

Reviewer: It's hard to predict because the number of inodes on an ext4 volume will depend on the volume size (see [this answer](https://unix.stackexchange.com/questions/785941/has-the-default-value-for-bytes-per-inode-changed)), and garage's filecount will depends on file size, configured block_size setting and object count.

For example uploading a 1GB video file will result (with the default 1MB block_size) in 1024 files on disk.

However we can make some napkin calculations: assuming ext4 defaults, for a volume below 4TB, we have 16,384 bytes per inode. So as long as file are bigger than 16k in their compressed state, you'll fill up the volume before running out of inodes.

* Metadata partition: on **SSD** (HDD OK if lots of RAM for kernel caching), **BTRFS or ZFS** (EXT4 OK with Garage-Snapshot)

> * Database: **LMDB is default, more tested, more performant. Recommended if Metadata on HDD**. (Prefer SQLite on a 32-bit system if you plan to store much data.)

Reviewer: ```suggestion
* Database: **LMDB is default, more tested, more performant assuming stable power and reliable storage**. Otherwise use SQLite. Always use SQLite on a 32 bit system.
```

> * Metadata may get corrupt, e.g. after a power outage: **Setup Garage-Snapshot** (SQLite should be avoided, but can be more robust with frequent unclean shutdown)

Reviewer: ```suggestion
* Metadata may get corrupt, e.g. after a power outage: **Setup Garage-Snapshot** (SQLite is more resilient to frequent unclean shutdown)
```

* (Use SQLite if you want to be able to migrate Metadata to a different architecture without resyncing, e.g. from AMD64 to ARM64.)

> * `blocksize = "10M"` if you have FTTH and plan to store mostly large files, **leave blocksize to default otherwise**

Reviewer: There are relatively few downsides to increasing the blocks size. Block size is actually a "max" block size, any file _above_ this limit will be cut in multiple files. For example if the block size is set to 1MB (default), then a 1.5MB file will be stored in two blocks of 1MB and 0.5MB. But if you set the block_size to 10MB, it will result in a single 1.5MB block. In practice, the blocks are compressed, so they will rarely reach the actual block size on disk.

The downsides of increasing the blocks size are increased memory usage and longer transfer time between nodes (RPC operations). They are relatively minor, so unless you operate in very limited memory and networking, raising the block size can benefit you greatly or even save you significant overhead, especially if your content is big (video or heavy images).

Author: what about changing the default value to 10MB upstream then ?

Author: I've set `10MB` default everywhere

Reviewer: > what about changing the default value to 10MB upstream then ?

It's something we can consider.

**Recommended** (minimal) data-center config:
* Data partition: on **SSD** (HDD OK if no high-performance storage), **XFS**, BTRFS & ZFS OK too
* Metadata partition: on **SSD**. **BTRFS or ZFS with Garage-Snapshot** (`ext4` with Garage-Snapshot)
* (You may replace Garage-Snapshot with filesystem snapshots if you stop garage during the snapshot)

> * Database: **LMDB** (SQLite if on 32-bit architecture)

Reviewer: No datacenter-like config should use 32 bit systems nowdays 😅

Author: yes, I'll remove this

* `blocksize = "10M"` if you plan to store mostly large files, **leave blocksize to default otherwise**

## Simple self-hosting Backups use-case

* Yunohost installed on a 60GB SSD (NVME or SATA)
* `/` is mounted on `ext4` partition `/dev/sda1` and so is `/home/yunohost.app/garage/`
* An empty 16TB HDD `/dev/sdb` (USB3 or SATA, USB2 should be avoided)
* Create a 16TB partition `/dev/sdb1` for the Data
* Type `/dev/sdb1` for the Data and `no` for the Metadata at install
* `/dev/sdb1` will be formatted in `xfs` and mounted on `/home/yunohost.app/garage/data`

> * Garage will take Snapshots of your LMDB database regularly

Reviewer: You might want to replace this with SQLite, especially if you have a single node

Author: is ext4 OK in this use-case or should we say metadata on ext4 is only OK for e.g. 4TB of data max ?

Author: I've set SQLite SQLite
